corehq/apps/linked_domain/keywords.py

```diff
@@ -10,24 +13,52 @@ def create_linked_keyword(domain_link, keyword_id):
     try:
         keyword = Keyword.objects.get(id=keyword_id, domain=domain_link.master_domain)
     except Keyword.DoesNotExist:
-        return
+        return None
 
     keyword_actions = keyword.keywordaction_set.all()
 
+    keyword.master_id = keyword.id
     keyword.id = None
     keyword.domain = domain_link.linked_domain
     keyword.couch_id = uuid.uuid4().hex
     keyword.save()
 
-    master_app_to_linked_app = {
-        doc["family_id"]: doc["_id"]
-        for doc in get_brief_app_docs_in_domain(domain_link.linked_domain)
-        if doc.get("family_id", None) is not None
-    }
-
     for keyword_action in keyword_actions:
-        keyword_action.pk = None
+        keyword_action.master_id = keyword_action.id
+        keyword_action.id = None
         keyword_action.keyword = keyword
         if keyword_action.app_id is not None:
-            keyword_action.app_id = master_app_to_linked_app[keyword_action.app_id]
+            keyword_action.app_id = get_master_app_to_linked_app(domain_link.linked_domain)[keyword_action.app_id]
         keyword_action.save()
+
+    return keyword.id
+
+
+def update_keyword(domain_link, keyword_id):
+    linked_keyword = Keyword.objects.get(id=keyword_id)
+    master_keyword = Keyword.objects.get(id=linked_keyword.master_id)
+
+    for prop in ['keyword', 'description', 'delimiter', 'override_open_sessions']:
+        setattr(linked_keyword, prop, getattr(master_keyword, prop))
+
+    linked_keyword.save()
+
+    for linked_keywordaction in linked_keyword.keywordaction_set.all():
+        master_keywordaction = KeywordAction.objects.get(id=linked_keywordaction.master_id)
+        for prop in ['action', 'recipient', 'message_content']:
+            setattr(linked_keywordaction, prop, getattr(master_keywordaction, prop))
+
+        app_id = get_master_app_to_linked_app(domain_link.linked_domain)[master_keywordaction.app_id]
+        if linked_keywordaction.app_id != app_id:
+            linked_keywordaction.app_id = app_id
+
+        linked_keywordaction.save()
+
+
+@quickcache(vary_on=['domain'], skip_arg=lambda _: settings.UNIT_TESTING, timeout=5 * 60)
+def get_master_app_to_linked_app(domain):
```

Comment:
This will be subject to the same issue as reports, where it doesn't support linked apps with multiple downstream apps, right? What if this function just discarded any apps that share the same family_id, would that make this fail more quickly/loudly? Or threw an exception to alert the user that this isn't a supported workflow?

Reply:
Made a note on https://dimagi-dev.atlassian.net/browse/USH-167, doesn't need to block this PR.